Implementing Prism.Autofac.Forms

ederbond opened the pull request:
@brianlagunas Fixed my implementation to use the latest preview version of Xamarin.Forms which is what Prism is built against.

@joacar I dont think I should provide a service key when registering the PageDialogService implementation since autofac could resolve its interface type when needed..

dnfclas commented:
Hi __@ederbond__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, DNFBOT;

dansiegel commented on Source/Xamarin/Prism.Autofac.Forms/AutofacExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+        /// <param name="container"></param>
+        public static void RegisterTypeForNavigation<T, C>(this IContainer container)
+            where T : Page
+            where C : class
```
For clarity this has been changed in Unity and Ninject to:
```cs
public static void RegisterTypeForNavigation<TView>( ... )
    where TView : Page

public static void RegisterTypeForNavigation<TView, TViewModel>( ... )
    where TView : Page
    where TViewModel : BindableBase
```

dansiegel commented on Source/Xamarin/Prism.Autofac.Forms/Prism.Autofac.Forms.csproj:
```diff
@@ -0,0 +1,94 @@
+    <RootNamespace>Prism.Autofac.Forms</RootNamespace>
+    <AssemblyName>Prism.Autofac.Forms</AssemblyName>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <TargetFrameworkProfile>Profile111</TargetFrameworkProfile>
```
Prism is currently targeting Profile 259 not 111.

joacar commented:
I think the reason for using service key is to ease overriding PageDialogService. I know AutoFac and DryIoc can handle this differntly (perhaps the service key is à must in unity/ninject). So service key is preferred for consistency. Please correct me if I'm wrong @brianlagunas

ederbond commented:
Hey @dansiegel, just updated to targe Profile 259 and updated the AutofacExtensions

`public static void RegisterTypeForNavigation<TView>( ... )
    where TView : Page

public static void RegisterTypeForNavigation<TView, TViewModel>( ... )
    where TView : Page
    where TViewModel : BindableBase`

dansiegel commented on Source/Xamarin/Prism.Autofac.Forms/AutofacExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+        /// <param name="container"></param>
+        public static void RegisterTypeForNavigation<TView, TViewModel>(this IContainer container)
+            where TView : Page
+            where TViewModel : class
```
should be ```where TViewModel : BindableBase```

ederbond replied:
@dansiegel Fixed

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/AutofacExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <param name="name">The name you will use to resolve the component in future.</param>
+        /// <param name="registerAsSingleton">Registers the type as a singleton.</param>
+        public static void RegisterTypeIfMissing(this IContainer container, Type type, string name, bool registerAsSingleton = false)
+        {
```
We don't use this method in XF, so it can be removed

ederbond replied:
Changed this to private since I use it twice on this class (Line 34 and Line 52).
I don't think I have a way to register a type without a ContainerBuilder on Autofac so I've created this method to help me registering new types on an existing container.

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/Modularity/AutofacModuleInitializer.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        readonly IComponentContext _context;
+        public AutofacModuleInitializer(IComponentContext context)
+        {
```
why does this not take IContainer like the PrismApplication calss?

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/Navigation/AutofacPageNavigationService.cs:
```diff
@@ -0,0 +1,24 @@
+        readonly IComponentContext _context;
+
+        public AutofacPageNavigationService(IComponentContext container, IApplicationProvider applicationProvider, ILoggerFacade logger)
+            : base(applicationProvider, logger)
```
why does this not take IContainer like the PrismApplication calss?

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/PrismApplication.cs:
```diff
@@ -0,0 +1,81 @@
+{
+    public abstract class PrismApplication : PrismApplicationBase<IContainer>
+    {
+        //const string _navigationServiceName = "AutofacPageNavigationService";
```
this needs to be implemented

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/PrismApplication.cs:
```diff
@@ -0,0 +1,81 @@
+
+        protected override INavigationService CreateNavigationService()
+        {
+            return Container.Resolve<INavigationService>();
```
need a named navigation service using the field defined on line 23

ederbond replied:
He @brianlagunas tecnically speaking Autofac doesn't need resolving instances based on names, it is able to resolve instances based on the implemented interface so this implementantion is more cleaner and will work perfectly. Of course I can register AutofacPageNavigationService withing a name and resolve it by name too but it is not a must have for Autofac.

brianlagunas replied:
I understand how IoC containers work, but the reason you need to do this is to support the extensibility of the INavigationService.  We provide a name to allow us to register an INavigationService without making it globally available through ctor injection.  In order to use it, it must be use by name, which hides it from the consumers of the framework, but allows the framework to use it internally.

ederbond replied:
Hummm, ok. I'll change that thanks for the patient to explain the reasons... :+1:

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/PrismApplication.cs:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        protected override void ConfigureContainer()
+        {
```
place the configuration in here, not in the CreateContainer.  CreateContiner should only return the container and nothing else.

brianlagunas commented:
We'll also need unit tests

ederbond commented:
Workin on unit tests... I'll providethem as soon as I have a break on my regular job.

brianlagunas commented:
Please rebase from Master. Fix compile issues (because I broke you), and fix tests.  Also, please update your extensions to follow the new pattern (see the UnityExtensions as an example).

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/Navigation/AutofacPageNavigationService.cs:
```diff
@@ -0,0 +1,28 @@
+        protected override Page CreatePage(string name)
+        {
+            if (!_container.IsRegisteredWithName<Page>(name))
+                throw new NullReferenceException($"The requested page 'missing' has not been registered.");
```
I think you were going for something more like:

`throw new NullReferenceException($"The requested page '{name}' has not been registered.");`

ederbond replied:
You're right I'm gonna fix that.

brianlagunas commented on README.md:
```diff
@@ -51,7 +51,7 @@ Following matrix shows the platform specific support currently available.
 | [Prism.Unity.Forms][10]| [![30]][10]|     |     |  X  |
 | [Prism.Mef][6]  <sup>(**)</sup> | [![26]][6] |  X  | - | - |
-| [Prism.Autofac][7]     | [![27]][7] |  X  |  X  |     |
+| [Prism.Autofac][7]     | [![27]][7] |  X  |  X  |  X  |
```
We will need a separate Prism.Autofac.Forms package and the PS build script updated to build the packages.

brianlagunas commented:
Looking good. Let's get it upgraded to the latest XF preview, and get the Prism.Autofac.Forms nupec created so we can get this merged into master.  Great job!

ederbond commented:
Just Updated to XF 2.3.1.111-pre2

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/PrismApplication.cs:
```diff
@@ -0,0 +1,125 @@
+            builder.Register(ctx => new AutofacModuleInitializer(Container)).As<IModuleInitializer>();
+            builder.Register(ctx => new EventAggregator()).As<IEventAggregator>();
+            builder.Register(ctx => new DependencyService()).As<IDependencyService>();
+            builder.Register(ctx => new PageDialogService(ctx.Resolve<IApplicationProvider>())).As<IPageDialogService>();
```
These should be singletons. Compare with the Unity project

ederbond replied:
INavigationService must be singleInstance too ?

brianlagunas replied:
No, not the nav service

brianlagunas commented on Source/Xamarin/Prism.Autofac.Forms/PrismApplication.cs:
```diff
@@ -0,0 +1,125 @@
+            return navigationService;
+        }
+
+        protected override void InitializeModules()
```
this method is no longer needed. See Unity project

ederbond replied:
`protected INavigationService CreateNavigationService(Page page)` removed.

ederbond replied:
`protected override void ConfigureContainer() ` changed to register types as singletons.

ederbond replied:
Hey Brian, After removing the method below form PrismApplication
`protected INavigationService CreateNavigationService(Page page)
        {
            var navigationService = CreateNavigationService();
            ((IPageAware)navigationService).Page = page;
            return navigationService;
        }`

one of my tests stoped working (the test below):
`[Fact]
        public void Navigate_ViewModelFactory_PageAware()
        {
            var app = new PrismApplicationMock();
            var view = new AutowireView();
            var viewModel = (AutowireViewModel)view.BindingContext;
            var pageAware = (IPageAware)viewModel.NavigationService;
            var navigationServicePage = app.CreateNavigationServiceForPage();
            Assert.IsType<AutowireView>(pageAware.Page);
            var navigatedPage = ((IPageAware)navigationServicePage).Page;
            Assert.IsType<AutowireView>(navigatedPage);
            Assert.Same(view, pageAware.Page);
            Assert.Same(pageAware.Page, navigatedPage);
        }`
should I remove this test or take the method CreateNavigationService(Page page) to the PrismApplication back ?
Note that on Prism.DryIoc.Forms this method still exists...

brianlagunas replied:
Just remove the test.  That method was added by the DryIoC package author just to test that scenario.

brianlagunas commented:
Thanks for the great PR!
